### PR TITLE
New back-navigation gesture using index-pinch when left hand faces head

### PR DESCRIPTION
**Pull request by elima**
Current gesture for navigating back uses a pinch geature between thumb and ring fingertip. This gesture is far too easy to trigger involuntarily and results in many usability issues (and user complains).

While we are still defining a comprehensive solution for all hand-interactions in Wolvic, this change serves as a short/mid term fix. It replaces the existing gesture by a new one: a normal pinch (thumb + index fingertip) with left hand facing the head. This is inspired by Oculus behavior, where a hamburger icon is shown when left hand is facing head. This specific icon is already generalized on the controllers to produce a navigate-back action.

A circled, hamburger menu icon is shown locked to the index's fingertip, when the left hand is facing the head. As the user performs a thumb+index pinch, the icon reduces its size up to 50%, when the pinch is closed and a navigate-back action is triggered.

It has been tested on Quest2 and Pico4 devices.

In the future, we could generalize this new floating+button+with+icon as a new widget that reacts to pinching with specific fingers, and it will probably work for the current Pointer widget, which would become a variant of it.

It fixes https://github.com/Igalia/wolvic/issues/617 for the time being.

### CODE REVIEW

File: app/src/openxr/cpp/OpenXRInputSource.cpp [hunk 1]
```diff
@@ -565,25 +571,74 @@ void OpenXRInputSource::EmulateControllerFromHand(device::RenderMode renderMode,
...
         }
 
-        memcpy(&jointTransforms[i], &transform, sizeof(vrb::Matrix));
+        jointTransforms[i] = transform;
+    }
+
+    // Check whether palm of left hand is facing the head, to show navigate-back button
+    bool leftPalmFacesHead = false;
+    if (mHandeness == OpenXRHandFlags::Left) {
+#if defined(OCULUSVR)
```
Comment (svillar) [inline app/src/openxr/cpp/OpenXRInputSource.cpp:587]: I know Pico does not properly fill the data for PALM joint but this implementation is not good for extensibility. Instead you should guard with ifdefs PICO's specific wrong behaviour and leave this code unguarded because is suitable for any other platform.

File: app/src/openxr/cpp/OpenXRInputSource.cpp [hunk 2]
```diff
@@ -565,25 +571,74 @@ void OpenXRInputSource::EmulateControllerFromHand(device::RenderMode renderMode,
...
+            vrb::Vector vector = vrb::Vector(0.0f, 0.0f, 1.0f);
+            vrb::Matrix palmMatrix = jointTransforms[XR_HAND_JOINT_WRIST_EXT];
+            vrb::Matrix headPalmMatrix = head.Inverse().PostMultiply(palmMatrix);
+            vrb::Vector diffVector = headPalmMatrix.MultiplyPosition(vector).Normalize();
+            leftPalmFacesHead = diffVector.z() >= kPalmHeadRotationZThreshold;
+        }
+#endif
+    }
+
+    // Scale joints according to their radius (for rendering)
```
Comment (svillar) [inline app/src/openxr/cpp/OpenXRInputSource.cpp:601]: hmm seems unrelated to the patch?

File: app/src/openxr/cpp/OpenXRInputSource.cpp [hunk 3]
```diff
@@ -565,25 +571,74 @@ void OpenXRInputSource::EmulateControllerFromHand(device::RenderMode renderMode,
...
+        }
     }
 
     delegate.SetHandJointLocations(mIndex, jointTransforms);
     delegate.SetAimEnabled(mIndex, mHasAimState);
+    delegate.SetLeftHandActionEnabled(mIndex, leftPalmFacesHead);
     delegate.SetMode(mIndex, ControllerMode::Hand);
     delegate.SetEnabled(mIndex, true);
 
+    // Select action
```
Comment (svillar) [inline app/src/openxr/cpp/OpenXRInputSource.cpp:617]: Why moving this here?

**Review by svillar — CHANGES_REQUESTED** [review]
Haven't tried it yet. For now I've some comments

**Comment by svillar** [comment]
BTW, this should very likely have a precursor commit, removing the previous gesture.

File: app/src/main/cpp/ControllerContainer.cpp [hunk 4]
```diff
@@ -233,6 +235,31 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
+#if !defined(OCULUSVR)
+        // Initialize left hand action button, which for now triggers back navigation
+        if (controller.leftHanded) {
+            TextureGLPtr texture = create->LoadTexture("menu.png");
+            assert(texture);
+            texture->SetTextureParameter(GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            texture->SetTextureParameter(GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            // texture->Bind();
+
+            const float width = 1.75f;
```
[the remark this answers is not in the document]
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:248]: Agree

File: app/src/main/cpp/ControllerContainer.cpp [hunk 5]
```diff
@@ -233,6 +235,31 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
         }
+
+#if !defined(OCULUSVR)
+        // Initialize left hand action button, which for now triggers back navigation
+        if (controller.leftHanded) {
+            TextureGLPtr texture = create->LoadTexture("menu.png");
+            assert(texture);
+            texture->SetTextureParameter(GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            texture->SetTextureParameter(GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            // texture->Bind();
```
[the remark this answers is not in the document]
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:246]: Right, will fix that.

File: app/src/openxr/cpp/OpenXRInputSource.cpp (hunk 3 above)
Comment (elima) [reply app/src/openxr/cpp/OpenXRInputSource.cpp:617]: Because now I need the pinch gesture for the navigate-back button, which happens when there is no aim state.

File: app/src/openxr/cpp/OpenXRInputSource.cpp (hunk 1 above)
Comment (elima) [reply app/src/openxr/cpp/OpenXRInputSource.cpp:587]: Right, I wrote *palm* in the comment but I'm actually using the wrist joint, because Pico doesn't provide the palm joint. I will make the comment consistent, and add guards to use wrist on Pico and Palm in the rest (except Oculus).

File: app/src/openxr/cpp/OpenXRInputSource.cpp (hunk 2 above)
Comment (elima) [reply app/src/openxr/cpp/OpenXRInputSource.cpp:601]: This scaling was part of the previous loop over joints but I had to move it here because to calculate whether palm faces head I need the joint transforms before scaling.

File: app/src/main/cpp/ControllerContainer.cpp [hunk 6]
```diff
@@ -233,6 +235,31 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
 
             controller.handMeshToggle->AddNode(transform);
         }
+
+#if !defined(OCULUSVR)
```
[the remark this answers is not in the document]
Comment (elima) [reply app/src/main/cpp/ControllerContainer.cpp:239]: Yes, makes sense.

**Comment by felipeerias** [comment]
The functionality looks very good to me. This is a good way to implement Back navigation.

I have a concern regarding the icon. The one provided by the Oculus system has a grey background colour, and goes from translucent to opaque as the user pinches their fingers.

Screenshot_20230503_115904


Screenshot_20230503_115933

However, our icon uses a transparent background which makes it harder to see. Then it becomes smaller when the user pinches their fingers.

Screenshot_20230503_114631

Screenshot_20230503_114704

I find that the Oculus approach is clearer. Do you think that it would make sense to change our implementation so it is similar to that one?

File: app/src/main/cpp/BrowserWorld.cpp [hunk 7]
```diff
@@ -531,6 +531,15 @@ BrowserWorld::State::UpdateControllers(bool& aRelayoutWidgets) {
     if (controller.modelToggle)
       controller.modelToggle->ToggleAll(controller.mode == ControllerMode::Device);
 
+    if (controller.leftHandActionEnabled && controller.handActionButtonTransform != nullptr) {
+      vrb::Matrix matrix = controller.handJointTransforms[10]->GetTransform();
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp]: Accessing index 10 is confusing. We only have a OpenXR hand tracking implementation so I guess you're accessing the location of XR_HAND_JOINT_INDEX_TIP_EXT ?

In any case we should make it independent of the SDK. Maybe we could define an enum with the 26 hand joints from the OpenXR spec and hope for the best. Another option would be to ask the device delegate maybe...

File: app/src/main/cpp/ControllerContainer.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (svillar) [reply app/src/main/cpp/ControllerContainer.cpp]: Still there :)

File: app/src/main/cpp/ControllerContainer.cpp [hunk 8]
```diff
@@ -233,6 +235,31 @@ void ControllerContainer::SetHandJointLocations(const int32_t aControllerIndex,
...
+            TextureGLPtr texture = create->LoadTexture("menu.png");
+            assert(texture);
+            texture->SetTextureParameter(GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+            texture->SetTextureParameter(GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+            // texture->Bind();
+
+            const float width = 1.75f;
+            const float aspect = (float)texture->GetWidth() / (float)texture->GetHeight();
+
+            QuadPtr icon = Quad::Create(create, width, width / aspect, nullptr);
```
Comment (svillar) [inline app/src/main/cpp/ControllerContainer.cpp]: This is a bit weird. Why do you compute the aspect? In the end you're passing width and height...

File: app/src/openxr/cpp/OpenXRInputSource.cpp [hunk 9]
```diff
@@ -584,25 +590,74 @@ void OpenXRInputSource::EmulateControllerFromHand(device::RenderMode renderMode,
...
+    // Check whether palm of left hand is facing the head, to show navigate-back button
+    bool leftPalmFacesHead = false;
+    if (mHandeness == OpenXRHandFlags::Left) {
+#if defined(OCULUSVR)
+        if (mHandJoints[XR_HAND_JOINT_PALM_EXT].locationFlags & XR_SPACE_LOCATION_POSITION_VALID_BIT)
+            leftPalmFacesHead = !mHasAimState;
+#else
+        if (mHandJoints[XR_HAND_JOINT_WRIST_EXT].locationFlags & XR_SPACE_LOCATION_POSITION_VALID_BIT) {
+            vrb::Vector vector = vrb::Vector(0.0f, 0.0f, 1.0f);
+            vrb::Matrix palmMatrix = jointTransforms[XR_HAND_JOINT_WRIST_EXT];
```
Comment (svillar) [inline app/src/openxr/cpp/OpenXRInputSource.cpp]: Why are you using the wrist joint instead of the palm joint? Seems weird because all the variables are related to palm.

**Review by svillar — CHANGES_REQUESTED** [review]

File: app/src/main/cpp/BrowserWorld.cpp [hunk 10]
```diff
@@ -531,6 +531,31 @@ BrowserWorld::State::UpdateControllers(bool& aRelayoutWidgets) {
     if (controller.modelToggle)
       controller.modelToggle->ToggleAll(controller.mode == ControllerMode::Device);
 
+    if (controller.leftHandActionEnabled && controller.handActionButtonTransform != nullptr) {
+      // Layout the button between the thumb and index fingertips
+      const int OPENXR_HAND_JOINT_INDEX_TIP = 10;
+      const int OPENXR_HAND_JOINT_THUMB_TIP = 5;
```
Comment (svillar) [inline app/src/main/cpp/BrowserWorld.cpp:537]: We should make it independent of the SDK. Maybe we could define an enum with the 26 hand joints from the OpenXR spec and hope for the best. Another option would be to ask the device delegate maybe...
